web/src/app/components/timeline/timeline.component.ts

```diff
@@ -0,0 +1,431 @@
+    3: '#2196f3',
+    4: '#d50000',
+    5: '#34a853',
+    6: '#d50000',
+    7: '#2196f3',
+    8: '#2196f3',
+    9: '#d50000',
+    10: '#2196f3',
+    11: '#2196f3',
+    12: '#d50000',
+    13: '#2196f3',
+    14: '#eee',
+    15: '#2196f3',
+    16: '#d50000',
+    17: '#eee',
+    18: '#34a853',
+    19: '#eee'
+  };
+
+  // Note that we use the non-null assertion operator ('!') on `this.svg`
+  // in order to reassure the compiler that it will never be null or undefined.
+  @ViewChild('timeline') private timelineContainer!: ElementRef;
+  @Input() private pushInfos!: step189_2020.IPushInfo[] | null;
+  private data: Item[] = [];
+  private svg!: d3SVG;
+  private x: d3.ScaleTime<number, number> = d3.scaleTime();
+  private height = 0;
+  private width = 0;
+  private numRows = 0;
+
+  /**
+   * Extracts the pushID, state, and start and end time for each push in
+   * pushInfos and inserts them into Item interfaces, which are collectively
+   * stored in an array.
+   *
+   * @param pushInfos Array of pushes for one push def
+   */
+  private static populateData(pushInfos: step189_2020.IPushInfo[] | null):
+    [Item[], number] {
+
+    if (!pushInfos) { return [[], 0]; }
+    const data: Item[] = [];
+
+    pushInfos.forEach(pushInfo => {
+      if (!pushInfo) { return; }
+      const states = pushInfo.stateInfo;
+      if (!states) { return; }
+      const statesStartTime = states[0].startTimeNsec;
+      if (!statesStartTime) { return; }
+      const statesEndTime = states[states.length - 1].startTimeNsec;
+      if (!statesEndTime) { return; }
+      const pushID = pushInfo.pushHandle;
+      if (!pushID) { return; }
+      const state = states[states.length - 1].state;
+      if (!state) { return; }
+
+      // Convert the start and end time values to seconds.
+      // The unary operator coerces the value to number type.
+      const startTime = +statesStartTime / TimelineComponent.NSEC_PER_MSEC;
+      const endTime = +statesEndTime / TimelineComponent.NSEC_PER_MSEC;
+
+      // Store data points as instances of TimelineBar interface.
+      data.push({
+        pushID,
+        state,
+        startTime,
+        endTime,
+        row: 0
+      } as Item);
+    });
+
+    // Assign a row value to each push representing their horizontal placement
+    // on the timeline. Each row index corresponds to one group.
+    const rowIndex = this.divideIntoRows(data);
+    return [data, rowIndex];
+  }
+
+  /**
+   * Divides the data into row by assigning a row index (corresponding
+   * to their row placement on the timeline) such that all intervals within a
+   * row do not overlap. The algorithm takes a greedy approach by sorting
+   * the row by increasing start time, sequentially picking the next interval,
+   * and removing all intervals it overlaps with. It thus fits as many intervals
+   * as possible in the first row, does so for each row until there are
+   * no more intervals left. The runtime is O(n * log(n)), which results in
```

Comment:
```suggestion
   * no more intervals left. The runtime is O(n * log(n)), which happens in
```

Reply:
Both should work equally well since the runtime is a direct result of the worst case scenario